### ISSUE TITLE
[Feature] ExTensor Operations - Matrix, shape, indexing operations

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ProjectOdyssey#3013]
## Summary
ExTensor needs additional operations for complete functionality.

## Files Affected
- `shared/core/extensor.mojo:23-28` - TODO list in docstring
- `examples/alexnet-cifar10/train.mojo:434,439` - Batch processing/slicing
- `shared/training/trainer_interface.mojo:392` - Tensor slicing
- `shared/training/__init__.mojo:417` - Batch iteration
- `tests/shared/core/test_shape.mojo:288-430` - Shape manipulation tests
- `tests/shared/training/test_training_loop.mojo:102,121,141` - Data loader integration

## Operations Needed
- **Matrix ops (#2717)**: matmul, transpose, dot, outer
- **Shape manipulation (#2718)**: reshape, squeeze, unsqueeze, concatenate, split, tile, repeat, broadcast_to, permute
- **Element-wise math (#2719)**: exp, log, sqrt, sin, cos, tanh
- **Statistical (#2720)**: var, std, median, percentile
- **Indexing (#2721)**: slicing, advanced indexing

## Acceptance Criteria
- [ ] Implement or document status of matrix operations
- [ ] Implement or document status of shape manipulation
- [ ] Implement or document status of element-wise math
- [ ] Implement or document status of statistical operations
- [ ] Implement slicing for batch processing
- [ ] Update TODO comments with resolution

## Previous Issues
Replaces closed #2717, #2718, #2719, #2720, #2721

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

Based on my comprehensive analysis of the codebase, here is the detailed implementation plan:

---

# Implementation Plan: Issue #3013 - ExTensor Operations

## Summary

This issue consolidates 5 previous closed issues (#2717-#2721) to implement/document the status of ExTensor operations for matrix ops, shape manipulation, element-wise math, statistical ops, and indexing/slicing. **The codebase analysis reveals that most operations already exist** but require:

1. **Documentation updates** - Remove outdated TODO comments in extensor.mojo
2. **Gap analysis** - Only `split()` function is truly missing
3. **Slicing improvements** - Enhance batch processing in training/data loading
4. **Test updates** - Update placeholder tests in test_shape.mojo

### Current Implementation Status

| Category | Status | Files | Notes |
|----------|--------|-------|-------|
| Matrix Ops (#2717) | ✅ **IMPLEMENTED** | `shared/core/matrix.mojo` | matmul, transpose, dot, outer, inner, tensordot |
| Shape Manipulation (#2718) | ✅ **MOSTLY IMPLEMENTED** | `shared/core/shape.mojo` | reshape, squeeze, unsqueeze, concatenate, stack, tile, repeat, broadcast_to, permute. **Missing: `split()`** |
| Element-wise Math (#2719) | ✅ **IMPLEMENTED** | `shared/core/elementwise.mojo` | exp, log, sqrt, sin, cos, tanh, ceil, floor, round, abs, sign |
| Statistical Ops (#2720) | ✅ **IMPLEMENTED** | `shared/core/reduction.mojo` | variance, std, median, percentile (with backwards) |
| Indexing/Slicing (#2721) | ⚠️ **PARTIAL** | `shared/core/extensor.mojo` | Basic slicing works. Needs integration with training loop |

---

## Step-by-Step Implementation Tasks

### Phase 1: Documentation Updates (Priority: HIGH)

**Task 1.1: Update ExTensor Docstring TODO Comments**
- **File**: `shared/core/extensor.mojo:23-28`
- **Action**: Update Array API Categories to reflect actual implementation status
- **Change**: Replace `(TODO(#27xx))` markers with `✓` for implemented features

**Task 1.2: Update Training Integration TODOs**
- **File**: `examples/alexnet-cifar10/train.mojo:434,439`
- **Action**: Update comments - slicing IS implemented, real blocker is data loader integration

**Task 1.3: Update trainer_interface.mojo TODO**
- **File**: `shared/training/trainer_interface.mojo:392`
- **Action**: Document that tensor slicing exists, actual blocker is Python data loader

**Task 1.4: Update training __init__.mojo TODO**
- **File**: `shared/training/__init__.mojo:417`
- **Action**: Clarify the actual blocker (Python/Mojo interop, not slicing)

---

### Phase 2: Implement Missing `split()` Function (Priority: MEDIUM)

**Task 2.1: Implement split() in shape.mojo**
- **File**: `shared/core/shape.mojo` (add after line 569)
- **Description**: Split tensor into multiple parts along an axis
- **Signatures**:
  ```mojo
  fn split(tensor: ExTensor, num_splits: Int, axis: Int = 0) raises -> List[ExTensor]
  fn split_indices(tensor: ExTensor, indices: List[Int], axis: Int = 0) raises -> List[ExTensor]
  ```

**Task 2.2: Add split() to exports**
- **File**: `shared/core/__init__.mojo`
- **Action**: Add split, split_indices to shape module exports

---

### Phase 3: Update Placeholder Tests (Priority: MEDIUM)

| Task | File:Lines | Action |
|------|------------|--------|
| 3.1 | `test_shape.mojo:288-294` | Enable test_split_equal |
| 3.2 | `test_shape.mojo:297-308` | Enable test_split_unequal |
| 3.3 | `test_shape.mojo:316-325` | Enable test_tile_1d |
| 3.4 | `test_shape.mojo:328-338` | Enable test_tile_multidim |
| 3.5 | `test_shape.mojo:346-355` | Enable test_repeat_elements |
| 3.6 | `test_shape.mojo:358-368` | Enable test_repeat_axis |
| 3.7 | `test_shape.mojo:376-389` | Enable test_broadcast_to_compatible |
| 3.8 | `test_shape.mojo:423-434` | Enable test_permute_axes |

---

### Phase 4: Batch Processing Integration (Priority: HIGH)

**Task 4.1: Verify extract_batch functions work with slicing**
- **Action**: Ensure extract_batch and extract_batch_pair use ExTensor.slice()
- **Test**: `tests/shared/core/test_slicing.mojo` already has batch extraction tests

**Task 4.2: Update DataLoader to use native slicing**
- **File**: `shared/training/trainer_interface.mojo:385-396`
- **Action**: Replace placeholder batch creation with actual tensor slicing

---

### Phase 5: Comprehensive Testing (Priority: HIGH)

| Test Suite | Command | Validates |
|------------|---------|-----------|
| Matrix | `pixi run mojo test tests/shared/core/test_matrix.mojo` | matmul, transpose, dot, outer |
| Shape | `pixi run mojo test tests/shared/core/test_shape.mojo` | reshape, squeeze, split, tile, etc. |
| Elementwise | `pixi run mojo test tests/shared/core/test_elementwise.mojo` | exp, log, sqrt, sin, cos, tanh |
| Reduction | `pixi run mojo test tests/shared/core/test_reduction.mojo` | variance, std, median, percentile |
| Slicing | `pixi run mojo test tests/shared/core/test_slicing.mojo` | slice(), __getitem__ |

---

## Files to Modify/Create

### Files to Modify

| File | Changes | Lines |
|------|---------|-------|
| `shared/core/extensor.mojo` | Update TODO comments in docstring | 23-28 |
| `shared/core/shape.mojo` | Add split(), split_indices() functions | ~570 (insert) |
| `shared/core/__init__.mojo` | Export split functions | TBD |
| `examples/alexnet-cifar10/train.mojo` | Update/remove TODO comments | 434, 439 |
| `shared/training/trainer_interface.mojo` | Update TODO, implement batch slicing | 392 |
| `shared/training/__init__.mojo` | Update TODO comment | 417 |
| `tests/shared/core/test_shape.mojo` | Enable placeholder tests | 288-434 |

### Files to Create

None - all functionality exists or can be added to existing files.

---

## Testing Approach

### Unit Tests

1. **Matrix Operations** - Tests exist in `tests/shared/core/test_matrix.mojo`
2. **Shape Manipulation** - Tests in `tests/shared/core/test_shape.mojo` (enable placeholders)
3. **Element-wise Math** - Tests in `tests/shared/core/test_elementwise.mojo`
4. **Statistical Operations** - Tests in `tests/shared/core/test_reduction.mojo`
5. **Slicing/Indexing** - Tests in `tests/shared/core/test_slicing.mojo`

### Integration Tests

- Test DataLoader with real tensor slicing: `tests/shared/training/test_training_loop.mojo`
- Full model training: `tests/models/test_lenet5_e2e.mojo` (weekly)

---

## Success Criteria

### Acceptance Criteria from Issue

- [x] ✅ Implement or document status of matrix operations - **Already implemented in matrix.mojo**
- [x] ✅ Implement or document status of shape manipulation - **Mostly implemented, split() missing**
- [x] ✅ Implement or document status of element-wise math - **Already implemented in elementwise.mojo**
- [x] ✅ Implement or document status of statistical operations - **Already implemented in reduction.mojo**
- [ ] ⏳ Implement slicing for batch processing - **Slicing exists, integration needed**
- [ ] ⏳ Update TODO comments with resolution - **Phase 1 tasks**

### Verification Checklist

- [ ] All TODO comments referencing #2717-#2721 are updated
- [ ] `split()` function implemented and tested
- [ ] Placeholder tests in test_shape.mojo enabled and passing
- [ ] DataLoader uses native tensor slicing
- [ ] All unit tests pass
- [ ] Pre-commit hooks pass: `just pre-commit-all`
- [ ] CI validation passes: `just validate`

---

## Resource Usage

### Estimated Effort

| Phase | Tasks | Estimated Time | Complexity |
|-------|-------|----------------|------------|
| Phase 1: Documentation | 4 | 1 hour | Low |
| Phase 2: Implement split() | 2 | 2 hours | Medium |
| Phase 3: Update Tests | 8 | 2 hours | Medium |
| Phase 4: Batch Integration | 2 | 2 hours | Medium |
| Phase 5: Testing | 5 | 1 hour | Low |
| **Total** | **21** | **8 hours** | |

### Tool Calls Used

- Read: 14
- Grep: 10
- Glob: 2
- Write: 1 (attempted)
- **Total**: 27 tool calls

### Dependencies

- None - all required infrastructure exists
- Can be implemented independently without blocking other work

### Risk Assessment

| Risk | Probability | Impact | Mitigation |
|------|-------------|--------|------------|
| split() implementation complexity | Low | Low | Use existing slice() infrastructure |
| Test failures in existing code | Low | Medium | Run tests incrementally |
| DataLoader integration issues | Medium | Low | Python/Mojo interop is separate concern |

---

## Key Findings

### Operations Already Implemented

1. **Matrix Operations** (`matrix.mojo`): matmul (all variants), transpose, dot, outer, inner, tensordot
2. **Shape Manipulation** (`shape.mojo`): reshape, squeeze, unsqueeze, concatenate, stack, tile, repeat, broadcast_to, permute
3. **Element-wise Math** (`elementwise.mojo`): exp, log, sqrt, sin, cos, tanh, ceil, floor, round, abs, sign
4. **Statistical Ops** (`reduction.mojo`): variance, std, median, percentile (all with backwards)
5. **Slicing** (`extensor.mojo`): slice(), __getitem__ for 1D/multi-dimensional slicing

### Actual Blocker for Batch Processing

The real issue is NOT missing slicing - it's Python/Mojo data loader integration. The TODO comments reference #2721 (slicing) but the actual blocker is:
1. `run_epoch()` takes a `PythonObject` data loader
2. Python → Mojo tensor conversion is incomplete
3. Need native Mojo data loader or proper interop

**Comment by mvillmow** [comment]
✅ Committed changes to branch: 3013--feature--extensor-operations-

---
*Automated by implement_issues.py at 2025-12-31 10:03:29*
